mitodl/open-discussions
# Working availability facet based on course run dates

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #2123 

#### What's this PR do?
- Adds `earliest_start` and `earliest_end` fields for course runs
- Adds reverse nested aggregations for availability facets to Elasticsearch requests
- Makes the availability facet work again
- Filters out archived/prior runs from search results

#### How should this be manually tested?
- `docker-compose run web python manage.py recreate_index`
- Import some courses and bootcamps if you don't have any
  - `docker-compose run web python manage.py backpopulate_bootcamp_data`
  - `docker-compose run web python manage.py backpopulate_edx_data`
  - `docker-compose run web python manage.py backpopulate_ocw_data` (this will take hours, so you can terminate it after a few are imported)
- Go to `../courses/search`.  The availability facet should have the following categories in them, and checking/unchecking them should work as expected:
  - Available Now
  - In the next 30 days
  - In the next 60 days
  - In the next 90 days
  - Over 90 days

#### Any background context you want to provide?
The imported data for EdX in particular is very inconsistent about which date fields are populated for any given course run: `start_date`, `end_date`, `enrollment_start`, `enrollment_end`, `semester`, `year`.  So the `earliest_start` and `earliest_end` fields were created to handle this, and calculated as follows: the enrollment date if not null, else the start/end date if not null, else a rough date based on semester & year if not null, else null. 

#### Screenshots (if appropriate)
<img width="338" alt="Screen Shot 2019-08-27 at 9 15 09 AM" src="https://user-images.githubusercontent.com/187676/63775424-49871e00-c8ad-11e9-88e9-98dd697f4da6.png">



#### What GIF best describes this PR or how it makes you feel?
![NQqf](https://user-images.githubusercontent.com/187676/63708220-d379ad00-c801-11e9-97f0-c71b589cd0e5.gif)
